zebra-state/src/request.rs

```diff
@@ -257,41 +257,54 @@ impl Treestate {
 /// database and recompute a new one.
 pub struct ContextuallyVerifiedBlockWithTrees {
     /// A block ready to be committed.
-    pub block: SemanticallyVerifiedBlock,
+    pub verified: SemanticallyVerifiedBlock,
```

Comment:
Should this type or the outer type name be different? Currently they don't match, which was part of your original feedback on the rename PR.

```suggestion
    pub verified: ContextuallyVerifiedBlock,
```

Reply:
Oh, thanks for catching this. You're right.

However, after looking into it again, I realized `ContextualyVerifiedBlock`s have two extra fields compared to `SemanticallyVerifiedBlock`s which aren't used anywhere in `ContextuallyVerifiedBlockWithTrees`. Why do we even call the struct `ContextuallyVerifiedBlockWithTrees`?

We construct `ContextuallyVerifiedBlockWithTrees` only in one place and either directly from a `ContextuallyVerifiedBlock` or `CheckpointVerifiedBlock`. I could simply construct a `ContextuallyVerifiedBlock` from a `CheckpointVerifiedBlock` with the two aforementioned fields being empty, but couldn't we rename `ContextuallyVerifiedBlockWithTrees` to `SemanticallyVerifiedBlockWithTrees` instead? It would generalize both the `CheckpointVerifiedBlock` and `ContextuallyVerifiedBlock`, while having all the fields necessary for committing the block.

Reply:
Because of the way we've defined different verification stages, a checkpoint verified block needs to have its header bound to the authorising data in its transactions. We also bind the header to the chain history of the previous block, but that's redundant verification. (We do it so light clients can use the history tree to quickly check we haven't left out any transactions in RPC responses.)

However, a block from the non-finalised state doesn't need either of these verification steps, because the non-finalised state has already contextually verified it, and generated its trees.

So I suggest we use names and types that reflect the verification status of the block, then adjust the data to match those names. If we have to add new names, types, or data, that's ok.

I don't think we should be constructing contextually verified blocks from checkpoint verified blocks, because that misrepresents their verification status. (Unless we've actually done the required contextual verification first!)

Reply:
It seems that with the current state of the renames the best thing to do here is to change the outer type from `ContextuallyVerifiedBlockWithTrees` to `SemanticallyVerifiedBlockWithTrees`.

This rename will work for the checkpointing because `SemanticallyVerifiedBlock` is enough to commit and for write a finalized block into the database because the block is already contextually verified : https://github.com/ZcashFoundation/zebra/blob/add-FinalizableBlock/zebra-state/src/service/write.rs#L293

Reply:
@oxarbitrage I thought of doing the same, but given Teor's comment, I'm not sure about this solution.

> Are we constructing contextually verified blocks from checkpoint verified blocks after doing the required verification? If we're not, that misrepresents their verification status.

The only two places where we construct `SemanticallyVerifiedBlockWithTrees` is 
1. https://github.com/zcashfoundation/zebra/blob/67d880d7f03a5758727de11d342faf5aedb671d5/zebra-state/src/service/finalized_state.rs#L283, and
2. https://github.com/zcashfoundation/zebra/blob/67d880d7f03a5758727de11d342faf5aedb671d5/zebra-state/src/service/finalized_state.rs#L298.

Both are from the new `FinalizableBlock`, which carries either a `CheckpointVerifiedBlock`, or `ContextuallyVerifiedBlock`. So what about calling it `FinalizableWithTrees`?

Reply:
My apologies, i touched this PR only because i had the information that you were going to be out the whole week. 

Feel free to reverse any of changes i did.

Reply:
The structure would look like this 
![image](https://github.com/ZcashFoundation/zebra/assets/8863307/199e4184-f3a2-409a-a1fc-372336b6a975) 

I don't like that `FinalizableWithTrees` would contain `SemanticallyVerified` again, though.

Reply:
@oxarbitrage, no bother. I want to finish it and don't leave it dangling.

Reply:
> I don't like that FinalizableWithTrees would contain SemanticallyVerified again, though.

We could create another wrapper of `SemanticallyVerified`, call it just `Verified`, and comment in the docs that it's either a former `CheckpointVerified`, or `ContextuallyVerified` block?

Reply:
What about `FinalizableBlockWithTrees`, and containing a `FinalizableBlock`?

I think we can merge what we have for now, and then open a new ticket and wait for Pili to schedule it?

Reply:
> What about `FinalizableBlockWithTrees`, and containing a `FinalizableBlock`?

The issue is that [`FinalizableBlock`](https://github.com/zcashfoundation/zebra/blob/7e7ce2ba712e07971d96ef4c44b8af9092041ae0/zebra-state/src/request.rs#L269) contains a `Treestate` in its `Contextual` variant, so `FinalizableBlockWithTrees` would contain the `Treestate` twice if `FinalizableBlock` was of that variant.

I'll think of a suitable solution, and if I can't come up with anything, I'll open a ticket.

If you have any further ideas, please share.

Reply:
I think I assumed that we'd remove the duplicate tree field from either `FinalizableBlockWithTrees` or `FinalizableBlock`.

But I can't keep those structs in my head, so listing them in a ticket might help.